components/movementsensor/wheeledodometry/wheeled_odometry.go

```diff
@@ -0,0 +1,341 @@
+}
+
+func (o *odometry) Orientation(ctx context.Context, extra map[string]interface{}) (spatialmath.Orientation, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	ov := &spatialmath.OrientationVector{Theta: o.orientation.Yaw, OX: 0, OY: 0, OZ: 1}
+	return ov, nil
+}
+
+func (o *odometry) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.orientation.Yaw, nil
+}
+
+func (o *odometry) LinearVelocity(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.linearVelocity, nil
+}
+
+func (o *odometry) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return geo.NewPoint(o.position.X, o.position.Y), o.position.Z, nil
+}
+
+func (o *odometry) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	return movementsensor.Readings(ctx, o, extra)
+}
+
+func (o *odometry) Accuracy(ctx context.Context, extra map[string]interface{}) (map[string]float32, error) {
+	return map[string]float32{}, movementsensor.ErrMethodUnimplementedAccuracy
+}
+
+func (o *odometry) Properties(ctx context.Context, extra map[string]interface{}) (*movementsensor.Properties, error) {
+	return &movementsensor.Properties{
+		LinearVelocitySupported:     true,
+		AngularVelocitySupported:    true,
+		OrientationSupported:        true,
+		PositionSupported:           true,
+		LinearAccelerationSupported: false,
+	}, nil
+}
+
+func (o *odometry) Close(ctx context.Context) error {
+	o.cancelFunc()
+	o.activeBackgroundWorkers.Wait()
+	return nil
+}
+
+// trackPosition uses the motor positions to calculate an estimation of the position, orientation,
+// linear velocity, and angular velocity of the wheeled base.
+// The estimations in this function are based on the math outlined in this article:
+// https://stuff.mit.edu/afs/athena/course/6/6.186/OldFiles/2005/doc/odomtutorial/odomtutorial.pdf
+func (o *odometry) trackPosition(ctx context.Context) {
+	o.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer o.activeBackgroundWorkers.Done()
+		ticker := time.NewTicker(time.Duration(o.timeIntervalMSecs) * time.Millisecond)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// Sleep until it's time to update the position again.
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+			// always use the first pair until more than one pair of motors is supported in this model
+			leftMove, err := o.motors[0].left.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+				return
+			}
+			rightMove, err := o.motors[0].right.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+				return
+			}
+			if !leftMove && !rightMove {
+				// neither motors are moving, so the odometry values do not need to be updated
```

Comment:
Suppose they were moving last time we checked, and this is the first time they're no longer moving. Should we update the values, since the motors have moved since the last time we updated? I wonder if the code would work smoother if we always updated no matter what, and removed this check entirely.

Reply:
maybe! I'll remove it and mess with the tests and make sure it doesn't cause any issues. originally it was a problem with the math adding even when the motors weren't moving but I think I eliminated that problem